core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -228,29 +212,33 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
     // if the operation is completed (by another thread) between the two tryComplete() calls, the
     // operation is unnecessarily added for watch. However, this is a less severe issue since the
     // expire reaper will clean it up periodically.
-
-    // At this point the only thread that can attempt this operation is this current thread
-    // Hence it is safe to tryComplete() without a lock
-    var isCompletedByMe = operation.tryComplete()
-    if (isCompletedByMe)
-      return true
-
-    var watchCreated = false
-    for(key <- watchKeys) {
-      // If the operation is already completed, stop adding it to the rest of the watcher list.
-      if (operation.isCompleted)
-        return false
-      watchForOperation(key, operation)
-
-      if (!watchCreated) {
-        watchCreated = true
-        estimatedTotalOperations.incrementAndGet()
-      }
-    }
-
-    isCompletedByMe = operation.maybeTryComplete()
-    if (isCompletedByMe)
-      return true
+    //
+    // ==============[story about lock]==============
+    // There is a potential deadlock between the callers to tryCompleteElseWatch() and checkAndComplete() in practice
+    // if we don't hold the lock while adding the operation to watch
+    // list and do the final tryComplete() check. For example,
```

Comment:
Perhaps change the above to the following?

We hold the operation's lock while adding the operation to watch list and doing the tryComplete() check. This is to avoid a potential deadlock between the callers to tryCompleteElseWatch() and checkAndComplete(). For example, the following deadlock can happen if the lock is only held for the final tryComplete() check.